src/runtime/security/negotiation_service.cpp

```diff
@@ -0,0 +1,51 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "negotiation_service.h"
+#include "negotiation_utils.h"
+#include "server_negotiation.h"
+
+namespace dsn {
+namespace security {
+extern bool FLAGS_enable_auth;
+
+negotiation_service::negotiation_service() : serverlet("negotiation_service") {}
+
+void negotiation_service::open_service()
+{
+    register_rpc_handler_with_rpc_holder(
+        RPC_NEGOTIATION, "Negotiation", &negotiation_service::on_negotiation_request);
+}
+
+void negotiation_service::on_negotiation_request(negotiation_rpc rpc)
+{
+    dassert(!rpc.dsn_request()->io_session->is_client(),
+            "only server session receive negotiation request");
+
+    // return SASL_AUTH_DISABLE if auth is not enable
+    if (!security::FLAGS_enable_auth) {
```

Comment:
negotiation_service is opened only if FLAGS_enable_auth is true, so it will never happen?

Reply:
If `FLAGS_enable_auth` is false,  it means server is not enable auth. So there is no need to do negotiation and return to client with `SASL_AUTH_DISABLE`. When client receives `SASL_AUTH_DISABLE`, it will set negotiation success

Reply:
In src/runtime/service_api_c.cpp:L556, `negotiation_service::open_service ` is called only when `FLAGS_enable_auth ` is true, so if program execute here, `FLAGS_enable_auth ` must be true, not needed to judge it.

Reply:
I changed the code in `service_api_c.cpp` like this:
```
service_app::service_app(const dsn::service_app_info *info) : _info(info), _started(false)
{
    security::negotiation_service::instance().open_service();
}
```
Remove the judgement of `if (FLAGS_enable_auth)`